Update test and contract collection docs (#2656)
foundry-rs/starknet-foundry: docs/src/testing/contract-collection/new-mechanism.md

# How Contracts Are Collected

> For the [`declare`](../../appendix/snforge-library/declare.md) to function, snforge must collect and call build on
> contracts in the package. By default, if using Scarb version greater or equal to 2.8.3, snforge will combine test
> collection and contract collection steps.

Reviewer: Not sure if it should be:
- ... function to work, snforge must collect...
- ... function, snforge must collect...

```suggestion
For the [`declare`](../../appendix/snforge-library/declare.md) function to, snforge must collect and call build on
contracts in the package. By default, if using Scarb >= 2.8.3, snforge will combine test
collection and contract collection steps.
```

Author: I meant "declare function", not "for declare to operate (function)".

Author: For the declare to function, ...

Author: But changed to `>=`.

When running `snforge test`, snforge will, under the hood, call the `scarb build --test` command. This command builds
all the test and contracts along them. Snforge collects these contracts and makes them available for declaring in tests.

Contracts are collected from both `src` and `tests` directory, including modules marked with `#[cfg(test)]`.
Internally, snforge collects contracts from all `[[test]]` targets compiled by Scarb.

> You can read more about that in [test collection](../test-collection.md) documentation.

Reviewer 2: Not clear what this sentence means. It suggests that you can read more about collecting contracts in the test collection documentation, which is not the case.

Author: I meant that test collection docs contain more details on `[[test]]` target. 
How about this?

```suggestion
You can read more about the `[[test]]` target in [test collection](../test-collection.md) documentation.
```

Reviewer 2: Sorry, I reviewed this before I saw the PR with the updated test collection docs. Everything makes sense now as it is.

## Collection Order

When multiple `[[test]]` targets are present, snforge will first try to collect contracts from `integration` `test-type`
target. If `integration` is not present, snforge will first collect contracts from the first encountered `[[test]]`
target.

> After collecting from initial `[[test]]` target, snforge will collect contracts from any other encountered contracts.

Reviewer: > snforge will collect contracts from any other encountered contracts.
should this sound like this 😅 ?

Author: Should have been `encountered targets`. Chaning.

No specific order of collection is guaranteed.

> 📝 **Note**
>
> If multiple contracts with the same name are present, snforge will use the first encountered implementation and will
> not collect others.

## Using External Contracts in Tests

> To use contract from dependencies in tests `Scarb.toml` must be updated to include these contracts under

Reviewer: nit:
```suggestion
To use contract from dependencies in tests, `Scarb.toml` must be updated to include these contracts under
```

`[[target.starknet-contract]]`.

```toml
[[target.starknet-contract]]
build-external-contracts = ["path::to::Contract1", "other::path::to::Contract2"]

> ```

Reviewer 2: I think a note with the link to the Scarb documentation should be added, the same as [here](https://github.com/foundry-rs/starknet-foundry/pull/2656/files#diff-2863525940dddfcea643e7feb3f5228c1003f6b9a3012b522f31558492c5ae61R30-R31)

Author: Added

# How Contracts Are Collected

> For the [`declare`](../../appendix/snforge-library/declare.md) to function, snforge must collect and call build on

Reviewer: ```suggestion
For the [`declare`](../../appendix/snforge-library/declare.md) to work, snforge must collect and call build on
```
